# 2016 Pascal Contest, Question 25

This notebook discusses the solution to question 25.

<img src="pascal-2016-question-25.pdf" width="100%">

## Representing grids in Python

We can represent a grid row as a triple of objects in left to right order.

In [1]:
row_123 = 1, 2, 3
print(row_123)

(1, 2, 3)


We represent a whole grid as a triple of rows in top to bottom order.

In [2]:
row_top = 1, 2, 3
row_middle = 4, 5, 6
row_bottom = 7, 8, 9
grid_tmb = row_top, row_middle, row_bottom
print(grid_tmb)

((1, 2, 3), (4, 5, 6), (7, 8, 9))


## Printing grids nicely

It will be useful to print a grid nicely, as a square.
Let's assume that the grid contains single digits or characters so they do form a square 
when printed.

In [3]:
def print_grid(grid):
    (a, b, c), (d, e, f), (g, h, i) = grid
    print(a, b, c)
    print(d, e, g)
    print(g, h, i)

grid_abc = ('a', 'b', 'c'), ('d', 'e', 'f'), ('g', 'h', 'i')
print_grid(grid_abc)

a b c
d e g
g h i


## Solving the problem by counting the grids using Python

The problem talks about rows and columns that contain 0's and 1's. We can represent these as Python tuples of integers. Since rows and column contain exactly three elements, we'll refer to them as triples. Let `triples` be the list of all possible triples that contain just 0's and 1's. Since each element of a triple can take two distinct values there are $2^3 = 8$ possible triples.

In [4]:
triples = [(x, y, z) for x in [0, 1] for y in [0, 1] for z in [0, 1]]
print(triples)
print(len(triples))

[(0, 0, 0), (0, 0, 1), (0, 1, 0), (0, 1, 1), (1, 0, 0), (1, 0, 1), (1, 1, 0), (1, 1, 1)]
8


The problem requires that each triple must contain at least one 0 and one 1. We'll refer to such a triple as a good triple. Let `good_triples` be the list of good triples.

In [5]:
good_triples = [t for t in triples if 0 in t and 1 in t]
print(good_triples)
print(len(good_triples))

[(0, 0, 1), (0, 1, 0), (0, 1, 1), (1, 0, 0), (1, 0, 1), (1, 1, 0)]
6


We can represent the grid as a tuple of three rows where each row is itself a triple of 0's and 1's. We can limit consideration to just rows that are good triples. Let `grids` be the list of all grids whose rows are good triples. There are three rows in each grid and each row can be one of the 6 good triples. Therefore, there are $6^3 = 216$ possible grids.

In [6]:
grids = [(row1, row2, row3) for row1 in good_triples for row2 in good_triples for row3 in good_triples]
print(len(grids))

216


A grid is good is each of its rows and columns is good. 
Let `is_good_grid` be a function that takes a grid as an input argument and returns `True` if and only if the grid is good.

In [7]:
def is_good_grid(grid):
    
    # check the rows
    if not grid in grids:
        return False
    
    # check the columns
    (a, b, c), (d, e, f), (g, h, i) = grid
    return \
        (a, d, g) in good_triples and \
        (b, e, h) in good_triples and \
        (c, f, i) in good_triples

grid_100 = (1, 0, 0), (0, 1, 0), (0, 0, 1)
print_grid(grid_100)
print(is_good_grid(grid_100), '\n')

grid_110 = (1, 1, 0), (0, 1, 0), (0, 0, 1)
print_grid(grid_110)
print(is_good_grid(grid_110), '\n')

grid_111 = (1, 1, 1), (0, 1, 0), (0, 0, 1)
print_grid(grid_111)
print(is_good_grid(grid_111))

1 0 0
0 1 0
0 0 1
True 

1 1 0
0 1 0
0 0 1
True 

1 1 1
0 1 0
0 0 1
False


Let `good_grids` be the list of good grids.

In [8]:
good_grids = [g for g in grids if is_good_grid(g)]

# print the first 3 good grids
for g in good_grids[:3]:
    print_grid(g)
    print()

0 0 1
0 0 1
1 1 0

0 0 1
0 1 1
1 0 0

0 0 1
0 1 1
1 0 1



The length of this list is the answer to the problem.

In [9]:
print(len(good_grids))

102


## Solving the problem mathematically

Now that we have computed the answer using Python, let's solve it using mathematical reasoning. We'll still use Python to help describe the mathematical concepts. 

Let $G$ be the number of good grids. This is what the problem is asking us to find. Above we observed that there are only 6 possibilities for the rows and therefore at most $6^3 = 216$ good grids. All of the candidate answers are less than 216 so we can't eliminate any of them immediately.

## Swapping 0s and 1s

Let's get more information about $G$. Observe that if we have one good grid, then we can easily find another, different, good grid by swapping 0s and and 1s. Therefore, every good grid has a twin. This means that $G$ must be a even number. 

We can focus on counting the members half the set of good grids.
The number of positions in the grid is 9 which is an odd number.
Therefore there can never be the same number of 0s and 1s.
Furthermore more, when we swap 0s and 1s, their counts swap.
This means we can focus on just the good grids that have less 0s than 1s.
For every such good grid its twin will have more 0s than 1s.
Let $G_0$ be the number of good grids that have less 0s than 1s.

$$
G = 2 G_0
$$

All the candidate solutions are even, so we still can't eliminate any of them. However, we now have simplified the problem a little since $G_0$ is half the size of $G$. So if we can compute $G_0$ then we know $G$. It might be easier to compute $G_0$.

## Swapping rows and swapping columns

Above we observed that swapping 0s and 1s transforms one good grid to another good grid.
Mathematicians refer to this as a *symmetry*.
There are other symmetries of the set of good grids. For example, if we have a good grid
then we can swap any pair of rows and get another good grid.
Similarly, we can swap any pair of columns.

Observe that since each row must have at least one 0, the total number of 0s must be at least 3, in which case there are 6 1s. Similarly there must be at least 3 1s and at most 6 0s. We can therefore count the solutions for each possible number of 0s, namely 3, 4, 5, and 6. Let $G_n$ be the number of good grids that have exactly $n$ 0s.

$$
G = G_3 + G_4 + G_5 + G_6
$$

Clearly

$$
G_0 = G_3 + G_4
$$

Now consider 0-1 swaps. If there are $n$ 0s in the grid, there must be $9-n$ 1s. So swapping 0 and 1 sends $n$ to $9-n$. Explicitly, the number of good grids that have 3 0s and 6 1s is the same as the number of good grids that have 6 0s and 3 1s, i.e. $G_3 = G_6$. Similarly $G_4 = G_5$. Therefore

$$
G = G_3 + G_4 + G_4 + G_3 = 2(G_3 + G_4) = 2 G_0
$$

Now let's consider $G_3$. Each row and each column must have a 0. Here's an example:

In [10]:
grid_3 = (0, 1, 1), (1, 0, 1), (1, 1, 0)
print_grid(grid_3)

0 1 1
1 0 1
1 1 0


Now consider the other good grids that have exactly 3 0s.
There are 3 possible columns for the 0 in row 1. For each of those, there are 2 possible columns for the 0 in row 2. This forces the column for the 0 in row 3. Therefore, there are 6 distinct good grids that have exactly 3 0s.

$$
G_3 = 6
$$

Now consider $G_4$. One row must have 2 0s and the other two rows must have 1 0.
Clearly, one of these rows must be in the column that does not contain the 0s in the doubled row. There are two cases to consider here according to where the final 0 goes, namely in the same column as the single 0 or in one of the columns of the doubled 0. 

Here is an example of case 1:

In [11]:
grid_41 = (0, 0, 1), (1, 1, 0), (1, 1, 0)
print_grid(grid_41)

0 0 1
1 1 1
1 1 0


Here is an example of case 2:

In [12]:
grid_42 = (0, 0, 1), (1, 1, 0), (1, 0, 1)
print_grid(grid_42)

0 0 1
1 1 1
1 0 1


Let $G_{41}$ and $G_{42}$ be the numbers of good grids for cases 1 and 2. 

$$
G_4 = G_{41} + G_{42}
$$

For case 1 we can find all the good grids by permuting the rows and colums of the example. There are only two distinct rows so the number of distinct row permutations is just 3. Similary, there are only 2 distinct columns so there are only 3 distinct column permutations.

$$
G_{41} = 3 \times 3 = 9
$$

Similarly, for case 2 there are 3 distinct rows and so 6 distinct row permuations. There are 3 distinct columns and therefore 6 distinct column permutations.

$$
G_{42} = 6 \times 6 = 36
$$

Therefore

$$
G_4 = G_{41} + G_{42} = 9 + 36 = 45
$$

and

$$
G = 2(G_3 + G_4) = 2(6 + 45) = 2 \times 51 = 102
$$

